### PR TITLE
feat(web): offer Live sight behind the native camera preview

### PR DESCRIPTION
**Pull request by Jasonnnz**
# Live sight on the native shells

The voice room's Live mode now works behind the Capacitor shells' native camera
preview. Three pieces:

1. The sight sampler's grid production (two-step area-averaged downscale to a
luma grid) moves behind `createFrameGridProducer()`; the browser sampler calls
the shared producer and its output is byte-identical (its suite passes
unmodified, and breaking the producer to point-sample reddens BOTH suites, so
the native path provably rides the same chain and one set of gate thresholds
serves both platforms).

2. `native-frame-source.ts` polls the camera plugin's `captureSample` (~1/s
while Live runs), decodes through that shared chain, and offers the grid to the
unchanged gate, handing back the exact judged JPEG with each decision. Runs are
generation-versioned so a capture or decode stranded by a stop/restart can
neither feed the old camera's frame to the new run nor suppress its ticks. The
module has no lifecycle opinion of its own: the owner stops it, through the
deduped `app.hidden` bus edge that also covers shells reporting a background
with no DOM event.

3. `use-voice-room-sight.ts` selects the source per viewfinder: the `<video>`
sampler in browsers, the native poller behind the preview layer. `liveAvailable`
drops its `!nativePreview` term; the preview being up is the capability answer,
since it exists only because the same plugin accepted a start. A keep reuses the
judged bytes (one round trip, and the transcript holds the frame the decision
was about) through the same guards, upload, send, and pulse as the browser path.
Photos and keeps encode at the same shared JPEG quality.

## Not in this PR

No haptics, no small-screen layout work, no QA-doc device rows (the plan's
polish PR), no Swift/Java changes, no daemon changes. `voice-room.tsx` is
untouched: the held-frame pulse already renders inside the native chrome
subtree.

## Verification

Per-file suites: use-voice-room-sight 69 (61 existing + 8 native),
native-frame-source 15, frame-sampler 22 (file unmodified), frame-gate 24 (file
unmodified), voice-room 123, camera-shutter 52, voice-camera 32, plus the sight
and HUD suites. Every new behavior assertion was red-green proved, including
both halves of the stale-run guard and the app.hidden chain (backgrounding
mid-poll refuses the in-flight frame via synchronous consent revocation).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: clients/web/src/lib/camera/native-frame-source.ts [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+
+/**
+ * Create a native frame source.
+ *
+ * Nothing is polled until {@link NativeFrameSource.start}, so a source can be
+ * built during render and started from an effect.
+ */
+export function createNativeFrameSource(
+  options: NativeFrameSourceOptions,
+): NativeFrameSource {
```
[the remark this answers is not in the document]
Comment (Jasonnnz) [reply clients/web/src/lib/camera/native-frame-source.ts:300]: Fixed in 8f480b156f by connecting it, as suggested. The source is now the native arm of the sight hook: the sampling effect in use-voice-room-sight.ts selects it whenever the native preview is up, liveAvailable drops its !nativePreview term, and a keep reuses the exact judged JPEG through the same guards, upload, and send as the browser path. Nothing in the module is orphaned; the hook suite grew 8 native-path tests covering source selection, mid-viewfinder swap, judged-bytes reuse, and quality pass-through.

File: clients/web/src/lib/camera/native-frame-source.ts [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+    running = false;
+    cancelPending();
+    document.removeEventListener("visibilitychange", handleVisibilityChange);
+  }
+
+  function start(): void {
+    stop();
+    running = true;
+    document.addEventListener("visibilitychange", handleVisibilityChange);
+    schedule();
```
[the remark this answers is not in the document]
Comment (Jasonnnz) [reply clients/web/src/lib/camera/native-frame-source.ts]: Fixed in 3a7ab8bbeb. The module now has no lifecycle handling at all: the visibilitychange listener, the hidden check, and the schedule/cancel pair are gone, and the docstring names the owner as the lifecycle authority. The owner already answers the deduped app.hidden bus edge (use-voice-room-sight.ts revokes capture consent synchronously and lowers the mode, whose effect stops the source), and the one poll that can land inside that commit is refused by the consent guard every capture path reads first. Covered by hook tests: backgrounding stops the poll, and a frame the poll offered after app.hidden is refused (no upload, no send, no pulse).

File: clients/web/src/lib/camera/native-frame-source.ts [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+    if (sampling) {
+      return;
+    }
+    sampling = true;
+    try {
+      const encoded = await captureSample();
+      // A stop, a flip, or a camera mid-teardown answers with nothing. That is
+      // the ordinary shape of this call and not a reason to end the poll.
+      if (!encoded || !running) {
+        return;
```
[the remark this answers is not in the document]
Comment (Jasonnnz) [reply clients/web/src/lib/camera/native-frame-source.ts]: Fixed in 3a7ab8bbeb with the suggested run generation. sampleOnce captures the generation at entry, every stop (and therefore start) bumps it, both post-await checks compare generations, and the in-flight claim is held per run so a sample stranded by a restart cannot suppress the new run's ticks. Both halves are test-covered: a capture and a decode from an ended run are each refused, and a run started over a stranded sample takes its first tick immediately.

**Comment by Jasonnnz** [comment]
@codex review this PR again — the previous issues have been fixed in commits 3a7ab8bbeb and 8f480b156f

File: clients/web/src/domains/chat/voice/voice-room/use-voice-room-sight.ts [hunk 4]
```diff
@@ -646,46 +687,78 @@ export function useVoiceRoomSight(
...
     videoRef,
   ]);
 
   // A flip points the camera somewhere else entirely and mirrors it, so every
   // score against the old baseline is meaningless and every capture still
-  // encoding belongs to the camera that is gone. The sampler keeps running: it
-  // is the same element, only the stream behind it changed.
+  // encoding belongs to the camera that is gone. The source keeps running: on
+  // the browser path it is the same element with a new stream behind it, and on
+  // the native path the poll never knew which camera it was reading.
```
[the remark this answers is not in the document]
Comment (Jasonnnz) [reply clients/web/src/domains/chat/voice/voice-room/use-voice-room-sight.ts]: Fixed in b5aea00264. The source gains invalidate() (a generation bump that leaves the cadence running), and the hook calls it from invalidateCaptures, which is the choke point both boundaries already pass through: the facing flip AND the transport reconnect, where the same in-flight hole existed and a stop/restart keyed to facing alone would have missed it. A capture or decode from before the boundary fails the generation compare and is refused before onDecision, so nothing reaches the upload path; the in-flight claim is per-run, so the stranded sample cannot suppress the next tick. Test-proved end to end at both layers: the hook suite pins that flips and reconnects tell the running poll (and that the browser path, whose capture reads the element at keep time, has nothing to refuse), and the source suite holds a capture and a decode across invalidate() and asserts zero offers plus an immediate next-tick flow.

**Comment by Jasonnnz** [comment]
@codex review this PR again — the previous issue has been fixed in commit b5aea00264

File: clients/web/src/lib/camera/native-frame-source.ts [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+    ? (encoded.match(/;base64,(.*)$/)?.[1] ?? "")
+    : encoded;
+  if (!base64) {
+    return null;
+  }
+  const binary = atob(base64);
+  const bytes = new Uint8Array(binary.length);
+  for (let index = 0; index < binary.length; index++) {
+    bytes[index] = binary.charCodeAt(index);
+  }
```
[the remark this answers is not in the document]
Comment (Jasonnnz) [reply clients/web/src/lib/camera/native-frame-source.ts]: Fixed in e1ec68560b. The decoder moved DOWN to utils/base64.ts (the documented home for pure helpers, importable by both lib/camera and domains/chat, which is why the sample path had a copy rather than an upward import), as decodeBase64Payload: data-URI vs bare normalization, the byte loop, and the empty/malformed contract in one place. All three prior copies are deleted in the same change, including the original dataUriToUint8Array in chat-attachments/utils.ts, so native photos, Live samples, and the pdf/text previews now share one decoder. Behavior is preserved at every call site, including the load-bearing null-vs-throw split (null = no bytes, use the other path; malformed = throw, so the PDF preview cannot hand broken bytes to pdf.js as a URL). The helper gained the suite none of the copies had, and each call site reddens on a different half of it, proving the extraction is live at all three.

**Comment by Jasonnnz** [comment]
@codex review this PR again — the previous issue has been fixed in commit e1ec68560b

File: clients/web/src/domains/chat/voice/voice-room/use-voice-room-sight.ts [hunk 6]
```diff
@@ -646,46 +709,81 @@ export function useVoiceRoomSight(
...
+      });
+      sampler.start(video);
+      stopSampling = sampler.stop;
+    } else {
+      const source = createNativeFrameSource({
+        gate,
+        captureSample: () =>
+          captureNativeVoiceCameraSample(NATIVE_CAPTURE_QUALITY),
+        onDecision: (decision, nowMs, sample) => {
+          recordFrameGateDecision("voice", decision, nowMs);
```
[the remark this answers is not in the document]
Comment (Jasonnnz) [reply clients/web/src/domains/chat/voice/voice-room/use-voice-room-sight.ts:734]: Fixed in 681806a04a. The HUD mount drops the !camera.native term (now just cameraOpen) and the stale comment is rewritten to present behavior: both viewfinders feed the same gate, and the readout is the instrument thresholds are tuned with on the hardware that runs them. Containment verified rather than assumed: the mount at voice-room.tsx:986 sits inside the motion.div carrying data-native-voice-camera-chrome (:796-:1522) with no portals in the file, and index.css keeps that subtree visible over the native preview. A new voice-room test pins the HUD mounting over a native preview with the readout enabled, red-green proved against exactly the reverted exclusion.

**Comment by Jasonnnz** [comment]
@codex review this PR again — the previous issue has been fixed in commit 681806a04a

File: clients/web/src/lib/camera/native-frame-source.ts [hunk 7]
```diff
@@ -0,0 +1,226 @@
...
+import { createFrameGridProducer, type FrameSource } from "./frame-sampler";
+
+/**
+ * Default gap between samples.
+ *
+ * Below the gate's own rate floor, so the gate and not this cadence decides how
+ * often a frame is kept, and far enough above a bridge round trip that a sample
+ * finishes long before the next tick asks for one.
+ */
+export const NATIVE_FRAME_SAMPLE_INTERVAL_MS = 1000;
```
[the remark this answers is not in the document]
Comment (Jasonnnz) [reply clients/web/src/lib/camera/native-frame-source.ts:112]: Fixed in e4950094d8 with paired sampling, the first of the suggested shapes. Each tick now takes two samples 60ms apart (half the motion window, leaving room for the second capture's own bridge+decode latency; derived from motionMaxAgeMs rather than written beside it): the first goes to a new record-only FrameGate.observe, which is exactly the remember-previous tail every offer already runs and provably nothing else (gate tests pin that an observed frame is never kept, moves no rate floor, no novelty baseline, no warmup, no settle grace), and the second is offered with real motion measured across the pair. The pan scenario is proved against the real gate with the shipped settle threshold: a panned pair skips as moving, a settled pair keeps, and deleting the observe call alone turns the pan into a keep. If the second capture fails, nothing is offered (falling back to the primer would reintroduce the blind keep), and a device too slow for the pairing degrades to exactly today's motion-absent behavior, visibly so in the readout. Cost is two bridge round trips per second while Live is held, noted in the module docstring for the device battery measurement.

**Comment by Jasonnnz** [comment]
@codex review this PR again — the previous issue has been fixed in commit e4950094d8

File: clients/web/src/lib/camera/native-frame-source.ts [hunk 8]
```diff
@@ -0,0 +1,319 @@
...
+      return grid ? { grid, blob } : null;
+    } finally {
+      frame.release();
+    }
+  }
+
+  /** Wait out the pair's spacing, on the timer the cadence already runs on. */
+  function waitPairSpacing(): Promise<void> {
+    return new Promise((resolve) => {
+      setTimeout(resolve, NATIVE_PAIR_SPACING_MS);
```
[the remark this answers is not in the document]
Comment (Jasonnnz) [reply clients/web/src/lib/camera/native-frame-source.ts]: Fixed in 7cb86b29ed, both halves. Stamps now come from capture resolution, not decode: each sample carries capturedAtMs taken the moment the bridge answers, observe/offer/onDecision all use those stamps, and the second capture is scheduled off the FIRST capture's stamp and fired without waiting for the primer's decode (the decode runs alongside it), so what separates the two stamps is the spacing plus one bridge call and decode latency is out of the measured gap entirely. And the invariant is enforced rather than hoped for: a pair whose capture-to-capture gap exceeds motionMaxAgeMs is DISCARDED, never offered, because a motion-null offer past the window is the blind keep this pairing exists to stop, so a device too slow to produce a comparable pair keeps nothing at all, visibly (Live runs and never pulses, the readout count stops advancing, and each discarded pair logs its measured gap). The slow-device signature and the number to report are in CAMERA_MODE_QA.md. The latency blind spot in the tests is closed: capture and decode stubs resolve on the fake timers with a stepped clock, and the suite pins capture-stamping under a 4x-window decode, discard for an out-of-window bridge, next-tick recovery, the one-millisecond-under fast path, and that the window constant tracks the shipped gate options (with an assertion that no slider can move motionMaxAgeMs).

**Comment by Jasonnnz** [comment]
@codex review this PR again — the previous issue has been fixed in commit 7cb86b29ed

File: clients/web/src/lib/camera/native-frame-source.ts [hunk 9]
```diff
@@ -0,0 +1,413 @@
...
+   * interval over a frame nobody wants.
+   */
+  let samplingGeneration: number | null = null;
+
+  /** Ask the bridge for one sample and stamp the moment it answers. */
+  async function captureStamped(run: number): Promise<CapturedSample | null> {
+    const encoded = await captureSample();
+    // The closest moment to the native snap that JS can observe, and the only
+    // one that keeps a decode out of the gap the gate measures.
+    const capturedAtMs = now();
```
[the remark this answers is not in the document]
Comment (Jasonnnz) [reply clients/web/src/lib/camera/native-frame-source.ts]: Fixed in db41d0a4a5. The pair's two stamps now take opposite ends of their captures' own uncertainty windows: the primer records its REQUEST time (the earliest its picture could have been taken) and the judged frame its answer (the latest), so the measured gap is an upper bound on true separation, a pair measuring inside the window is certainly inside it, and every error the bound makes is a discard rather than an accept. One nuance vs the example: moving the scheduling anchor to request time also means a slow primer spends the spacing budget inside its own bridge call, so the t=0/t=100/t=160 case now requests the second sample immediately on the primer's answer and measures ~100ms, which the 120ms window legitimately holds since the pictures truly cannot be further apart than that; a primer slow enough to actually break the bound is discarded. Both directions are pinned independently in tests (breaking only the primer's bound reddens only the first-capture cases and vice versa), and reverting to answer-time stamps reddens three tests including the undercount scenario.

File: clients/web/docs/CAMERA_MODE_QA.md [hunk 10]
```diff
@@ -68,21 +68,36 @@ scrim.
...
-above and these are never reachable in the same viewfinder. And in the installed
-app the web preview also appears as a fallback when a native start fails, which
-makes Live briefly available there; every later acquire retries the native
-start, so a flip that succeeds takes the preview back, and Live ends with it
-rather than the pill going on claiming a stream nothing is reading.
+### Live on iPhone
+
+Live runs behind either viewfinder. In mobile Safari it samples the room's own
+`<video>`; in the installed app, where the Capacitor plugin draws its preview
+behind the web view and there is no element to read, it polls the plugin for a
```
[the remark this answers is not in the document]
Comment (Jasonnnz) [reply clients/web/docs/CAMERA_MODE_QA.md:99]: Fixed in db41d0a4a5. The stale iPhone row (the native preview offers no Live = success) is replaced with rows asserting the hold offers Live behind the native preview and keeps pulse, pointing at the slow-bridge discard signature. A truth pass over the whole merged file also found the Android section had no Live coverage at all despite this PR enabling the identical poll there, so it gained a row referencing the iPhone Live checks; everything else in the file still holds.

**Comment by Jasonnnz** [comment]
@codex review this PR again — the previous issues have been fixed in commit db41d0a4a5 (the branch also carries a merge of main resolving the camera-view-settings conflict)

File: clients/web/src/lib/camera/native-frame-source.ts [hunk 11]
```diff
@@ -0,0 +1,442 @@
...
+  function waitUntil(targetMs: number): Promise<void> {
+    return new Promise((resolve) => {
+      setTimeout(resolve, Math.max(0, targetMs - now()));
+    });
+  }
+
+  async function sampleOnce(): Promise<void> {
+    const run = generation;
+    if (samplingGeneration === run) {
+      return;
```
[the remark this answers is not in the document]
Comment (Jasonnnz) [reply clients/web/src/lib/camera/native-frame-source.ts:484]: Fixed in 86af0275de, at MODULE scope rather than per source, which is the only scope matching the plugin's own single callback slot: an invalidate abandons a call still running on the bridge, and a source the room swaps out leaves one behind too, so any narrower queue would still put two calls on the plugin. Every captureSample now issues through one process-wide slot; generations keep deciding what a RESULT is worth. The earliest-bound stamp is read inside the queued work at actual issuance (stamping at enqueue would inflate the gap and discard back-to-back pairs), a refused call cannot wedge the queue (rejection handler on the slot, 22 tests redden without it), and a queued sample that waited behind a stranded call discards through the existing gap bound rather than falsely accepting. Concurrency is measured, not inferred: the capture stub tracks outstanding calls and asserts a high-water mark of one, including across a source the room replaced. A bridge that never answers at all stops the poll for good, deliberately: releasing the queue on a timeout would issue exactly the concurrent call the queue exists to prevent while the original is still outstanding natively, and the docstring states the resulting signature (Live keeps nothing, restart does not recover).

File: clients/web/src/domains/chat/voice/voice-room/use-voice-room-sight.test.tsx [hunk 12]
```diff
@@ -990,14 +1028,11 @@ describe("useVoiceRoomSight: sharing a keep", () => {
     await flush();
 
     // Availability going is the third way consent ends, and it ends it for the
-    // same reason: the preview the frame came from is not the one on screen.
+    // same reason: there is no longer anywhere honest for the frame to land.
```
[the remark this answers is not in the document]
Comment (Jasonnnz) [reply clients/web/src/domains/chat/voice/voice-room/use-voice-room-sight.test.tsx]: Fixed in 86af0275de: the comment now states the invariant directly (an unavailable Live has nowhere honest to land a frame). A scan of the PR's added lines for temporal phrasing found one more hit, a test name, which is also rewritten (says nothing to a stopped poll); the full-PR scan is at zero.

**Comment by Jasonnnz** [comment]
@codex review this PR again — the previous issues have been fixed in commit 86af0275de

File: clients/web/src/lib/camera/native-frame-source.ts [hunk 13]
```diff
@@ -0,0 +1,500 @@
...
+   * asked for it, which would inflate the pair's gap and discard a pair that
+   * was really taken back to back.
+   */
+  function issueCapture(): Promise<{
+    encoded: string | null;
+    requestedAtMs: number;
+  }> {
+    const issued = captureSlot.then(async () => {
+      const requestedAtMs = now();
+      const encoded = await captureSample();
```
[the remark this answers is not in the document]
Comment (Jasonnnz) [reply clients/web/src/lib/camera/native-frame-source.ts]: Fixed in fceb581706 as suggested: issueCapture takes the run and rechecks the generation inside the queued continuation at slot-open time, so a dead run resolves null without touching the bridge and the slot frees for whoever is next; stamp-at-issuance for live runs is unchanged. The docstring's wedge story is narrowed to match: only an ISSUED call can hold the queue. The new test walks the exact scenario (held call out, second run queues, that run stops while waiting, held call releases: the bridge saw one call, and a restart then samples normally through a free slot), and inverting the guard to drop live runs reddens 31 tests, so the check provably does not over-apply.

**Comment by Jasonnnz** [comment]
@codex review this PR again — the previous issue has been fixed in commit fceb581706

File: clients/web/src/domains/chat/voice/voice-room/use-voice-room-sight.ts [hunk 14]
```diff
@@ -360,19 +396,27 @@ export function useVoiceRoomSight(
...
+  // watching. Neither source stops itself: one pauses its loop while the page
+  // is hidden and picks it back up on the way in, the other polls a bridge that
+  // has no opinion about the app being on screen. A Live that survived a
+  // backgrounding would therefore resume sharing what the camera sees on a
+  // gesture made before the app was put away. Coming back to a viewfinder on
+  // photo costs one hold; coming back to one that is already streaming costs
+  // whatever it captured first.
+  //
+  // Lowering the mode is what stops the source, one commit later, and the
+  // revocation is what covers the commit: it is synchronous with the edge, and
```
[the remark this answers is not in the document]
Comment (Jasonnnz) [reply clients/web/src/domains/chat/voice/voice-room/use-voice-room-sight.ts]: Fixed in de155e1594. The comment now states the invariant directly: the revocation is synchronous with the edge, lowering the mode stops the source only when React's next commit runs the sampling effect's cleanup, and the revocation covers that window because every capture path reads it first.

**Comment by Jasonnnz** [comment]
@codex review this PR again — the previous issue has been fixed in commit de155e1594

File: clients/web/src/lib/camera/native-frame-source.ts [hunk 15]
```diff
@@ -0,0 +1,515 @@
...
+      return { encoded, requestedAtMs };
+    });
+    // The queue survives whatever this call does, so one refusal cannot wedge
+    // every later sample behind it.
+    captureSlot = issued.then(
+      () => undefined,
+      () => undefined,
+    );
```
[the remark this answers is not in the document]
Comment (Jasonnnz) [reply clients/web/src/lib/camera/native-frame-source.ts:414]: Fixed in fd77cb2f39. The slot now carries a last-resort release deadline (10s: two orders of magnitude past a healthy round trip, ten poll intervals, ~80x the motion window, so a call still outstanding could not produce a frame the gap bound would accept even if it landed), after which the waiting tick resolves empty and the queue advances, so the worst case is a poll quiet for at most ten seconds that resumes on its own, and Live restarted after the camera returns samples normally. Abandonment safety is verified against the installed plugin's native sources on BOTH platforms, not inferred: iOS stores a single sampleBufferCaptureCompletionBlock that a later call replaces and the delegate fires for the later call; Android overwrites snapshotCallbackId and onSnapshotTaken resolves whatever it names, so the overwrite strands only the abandoned promise nobody awaits and the new call gets the frame taken for it (Android's stale saved call lingers in the bridge map, a small bounded leak, noted). Red-green covers the tab-brick scenario (deadline release then a normal next run, with maxConcurrent()===2 asserted as the sanctioned exception so the suite documents the boundary), a call answering just under the deadline flowing untouched, and one abandonment disturbing neither a live run nor a decode in flight.

**Comment by Jasonnnz** [comment]
@codex review this PR again — the previous issue has been fixed in commit fd77cb2f39
